typescript/index.d.ts

```diff
@@ -116,10 +116,14 @@
  }
  export type AtomOptions<T> = AtomOptionsWithoutDefault<T> | AtomOptionsWithDefault<T>;
 
+ type DeepReadonly<T> = {
+  readonly [P in keyof T]: DeepReadonly<T[P]>
```

Comment:
Does this type accept arrays?

Reply:
Are you aware of any Flow equivalent?

Reply:
@mondaychen it does! Although I was missing a few special cases (`Map`,`Set`, functions) so I pushed an updated version. And @drarmstr I think in theory it would be possible but potentially painful since flow makes it hard to do conditional types, something like this should work (but seems to be an old gist, maybe there's a better way now): https://gist.github.com/karlhorky/99ecb289504dd96d1510ee3961aeeac0  , and it doesn't seem to have a special case for `Map` and `Set`

Reply:
Ah good question, in terms of custom classes it'll treat the properties as frozen, but methods that mutate the object are still allowed at the type level since we can't know ahead of time which methods in a custom class mutate the object. I think that might be an ok tradeoff though

Reply:
@drarmstr Ah good question, in terms of custom classes it'll treat the properties as frozen, but methods that mutate the object are still allowed at the type level since we can't know ahead of time which methods in a custom class mutate the object. I think that might be an ok tradeoff though

Reply:
@csantos42  Yeah, it's not perfect but seems like it could help.  The Flow team didn't have suggestions for a recommended helper for this.  I guess the primary concern is how much this might impact `dangerouslyAllowMutability` usage?

Reply:
good point @drarmstr , the experience for mutability should be taken into account so I overloaded the constructors to only make the value readonly when `dangerouslyAllowMutability` is false/not set. (and added corresponding unit tests)

Reply:
Nice!